src/execution/parsing/dag_parser.c

```diff
@@ -263,48 +268,90 @@ int ParseDAGRunCommand(RedisAI_RunInfo *rinfo, RedisModuleCtx *ctx, RedisModuleS
             if (dag_ro) {
                 RAI_SetError(rinfo->err, RAI_EDAGBUILDER,
                              "ERR PERSIST cannot be specified in a read-only DAG");
-                goto cleanup;
+                return REDISMODULE_ERR;
             }
             /* Store the keys to persist in persistTensors dict, these keys will
              * be mapped later to the indices in the dagSharedTensors array in which the
              * tensors to persist will be found by the end of the DAG run. */
             const int parse_result = _ParseDAGPersistArgs(&argv[arg_pos], argc - arg_pos,
-                                                          rinfo->persistTensors, "|>", rinfo->err);
+                                                          rinfo->persistTensors, rinfo->err);
             if (parse_result <= 0)
-                goto cleanup;
+                return REDISMODULE_ERR;
             arg_pos += parse_result;
             persist_complete = true;
             continue;
         }
+        if (!strcasecmp(arg_string, "KEYS") && !keys_complete && chainingOpCount == 0) {
+            const int parse_result = ParseKeysArgs(ctx, &argv[arg_pos], argc - arg_pos, rinfo->err);
+            if (parse_result <= 0)
+                return REDISMODULE_ERR;
+            arg_pos += parse_result;
+            keys_complete = true;
+            continue;
+        }
         if (!strcasecmp(arg_string, "TIMEOUT") && !timeout_complete && chainingOpCount == 0) {
             long long timeout;
             if (_parseTimeout(&argv[arg_pos], argc - arg_pos, &timeout, rinfo->err) ==
                 REDISMODULE_ERR)
-                goto cleanup;
+                return REDISMODULE_ERR;
             rinfo->timeout = timeout;
             arg_pos += 2;
             timeout_complete = true;
             continue;
         }
-
         if (!strcasecmp(arg_string, "|>") && arg_pos < argc - 1) {
-            RAI_DagOp *currentOp = _AddEmptyOp(&dag_ops);
+            RAI_DagOp *currentOp = _AddEmptyOp(dag_ops);
             chainingOpCount++;
             int args_num = _CollectOpArgs(argv, argc, ++arg_pos, currentOp);
             arg_pos += args_num;
             continue;
         }
         // If none of the cases match, we have an invalid op.
-        RAI_SetError(rinfo->err, RAI_EDAGBUILDER, "ERR Invalid DAGRUN command");
-        goto cleanup;
+        RedisModule_Log(ctx, "warning", "Command syntax error. Unexpected argument: %s",
+                        arg_string);
+        RAI_SetError(rinfo->err, RAI_EDAGBUILDER, "ERR Invalid DAG command");
+        return REDISMODULE_ERR;
     }
-
-    if (array_len(dag_ops) < 1) {
+    // This verification is needed for AI.DAGEXECUTE(_RO) commands (but not for the deprecated DAG
+    // commands).
+    if (!strncasecmp(RedisModule_StringPtrLen(argv[0], NULL), "AI.DAGEXECUTE",
```

Comment:
why is this check? you are here because you registered command and this is the code path. also, why not just use `strcasecmp` instead of `strncasecmp`?, also you can get the length from, `RedisModule_StringPtrLen`